### PR TITLE
stats: do not split excluded lower value ranges

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/11907

This bug would occur on indices that have multiply columns like `index idx(a,b)`, and the bug happens when:

- Queries that only request ranges on prefix like `where a >= 10` are issued and collected by feedback, which result in one of the bucket upper bound becomes single encoded value `(10)`, and there is a chance that next buckets's upper bound is `(10, 11)`.

- Then another queries comes like `where a >= 9`, so `SplitRange` is called with `[9, +inf)`, so we will split by `(10)` and `(10, 11)`, which result in ranges `[9, 10]`, `(10,(10,11)]`,..., and every thing looks fine now.

- The caller, which is `IndexRangesToKVRanges`, will process the splited ranges `(10,(10,11)]`. Since lower is excluded, `(10,(10,11)]` will be transformed to `[11, (10,12))` by using `PrefixNext`, so the invalid ranges happens.

### What is changed and how it works?

When we split ranges, do not generate execluded lower ranges. This PR does it by split the ranges by lower bound and always generate included lower ranges and excluded upper ranges.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

**Comment by lamxTyler** [comment]
@winoros Main reason that I did not choose this solution is that there are old histograms and it is not easy to determine the original number of columns, because it may already been `PrefixNext` many times and it may not decodeable. To use this approach we need to increase the stats version and only use feedback on newly created histograms.

File: statistics/histogram.go [hunk 1]
```diff
@@ -431,41 +431,43 @@ func (hg *Histogram) typeMatch(ranges []*ranger.Range) bool {
 	return true
 }
 
-// SplitRange splits the range according to the histogram upper bound. Note that we treat last bucket's upper bound
-// as inf, so all the split Ranges will totally fall in one of the (-inf, u(0)], (u(0), u(1)],...(u(n-3), u(n-2)],
-// (u(n-2), +inf), where n is the number of buckets, u(i) is the i-th bucket's upper bound.
+// SplitRange splits the range according to the histogram lower bound. Note that we treat first bucket's lower bound
+// as -inf and last bucket's upper bound as +inf, so all the split ranges will totally fall in one of the (-inf, l(1)),
+// [l(1), u(2)]),...[l(n-2), l(n-1)), [l(n-1), +inf), where n is the number of buckets, l(i) is the i-th bucket's lower bound.
```
Comment (eurekaka) [inline statistics/histogram.go]:
```suggestion
// [l(1), l(2)),...[l(n-2), l(n-1)), [l(n-1), +inf), where n is the number of buckets, l(i) is the i-th bucket's lower bound.
```

File: statistics/handle/update.go [hunk 2]
```diff
@@ -766,11 +766,11 @@ func formatBuckets(hg *statistics.Histogram, lowBkt, highBkt, idxCols int) strin
 		return hg.BucketToString(lowBkt, idxCols)
 	}
 	if lowBkt+1 == highBkt {
-		return fmt.Sprintf("%s, %s", hg.BucketToString(lowBkt, 0), hg.BucketToString(highBkt, 0))
+		return fmt.Sprintf("%s, %s", hg.BucketToString(lowBkt, idxCols), hg.BucketToString(highBkt, idxCols))
```
Comment (eurekaka) [inline statistics/handle/update.go]: Why do we need this change?

Comment (lamxTyler) [reply statistics/handle/update.go]: Without it, the result for index is unreadable.

**Pull request merged**